quickjs.c

```diff
@@ -8887,6 +8887,22 @@ static int JS_SetPropertyInternal2(JSContext *ctx, JSValue obj, JSAtom prop,
             pr = add_property(ctx, p, prop, JS_PROP_C_W_E);
             if (!pr)
                 goto fail;
+
+            // try give function name in every possible way
+            if (JS_IsFunction(ctx, val)) {
+                JSObject *vf = JS_VALUE_GET_OBJ(val);
+                if (vf->class_id == JS_CLASS_BYTECODE_FUNCTION) {
+                   const char* name = get_func_name(ctx, val);
+                    if (strcmp(name, "") == 0) {
```

Comment:
Minor style issues: alignment + star leans right; applies to the closing brace too

Correct me if I'm wrong but you want to set the name if it's *not* empty, right? Also, get_func_name can return NULL
```suggestion
                    const char *name = get_func_name(ctx, val);
                    if (name && *name) {
```

Reply:
> I don't think JS_SetPropertyInternal2 is the right place for this because then we're adding a property as a side effect of adding another property.

It's anonymous function by definition `function() {throw new Error('woof woof');};` It use the object property name as a fallback for easy debugging.

Reply:
> Correct me if I'm wrong but you want to set the name if it's not empty, right?

Yeap. quickjs will work when `dog.bark` value is named function
```js
var dog = {};
dog.bark = function c2() {
  throw new Error('woof woof');
};
function makeStack() {
  try {
    dog.bark();
  } catch (error) {
    print(error.stack)
  }
}
makeStack();
```

You can see it here http://k2.gengjiawen.com:8001/

Reply:
Yes, I understand that. I think in general it should be possible to recover that information from the JSStackFrame when the backtrace is built.

(I worked on something similar recently for constructors, although I haven't finished that yet.)

The problem I see with the approach in this PR is that it produces the wrong result in this example, right?
```js
var a = {foo: function() { throw Error("fail") }}
var b = {bar: a.foo}
b.bar() // "foo" in stack trace
```

Reply:
> Yes, I understand that. I think in general it should be possible to recover that information from the JSStackFrame when the backtrace is built.

Yeap. Is there enough context when construct backtrace since this request the property info ?

> The problem I see with the approach in this PR is that it produces the wrong result in this example, right?

Personally I think it's acceptable. the function is foo when defined.
v8 for reference
```console
/tmp/6bfa5ab:1: Error: fail
var a = {foo: function() { throw Error("fail") }}
                           ^
Error: fail
    at Object.foo [as bar] (/tmp/6bfa5ab:1:34)
    at /tmp/6bfa5ab:3:3
```

Reply:
> Also, get_func_name can return NULL

Is this why the segment fault in ASAN ?

Reply:
> Is there enough context when construct backtrace

The answer is a qualified "yes". I had it sort of working for constructors but in general no. I'll grant that making JS_CallInternal record that information is probably a lot more involved than your current approach.

That said, I still don't think JS_SetPropertyInternal2 is the proper place for that because that makes the .name property magically appear as an observable side effect of another operation.

After thinking about it some more, I think the right place is in the parser. We already parse something like this correctly:
```js
var f = function() {} // f.name === "f"
```
And also:
```js
var o = {f: function() {}} // o.f.name === "f"
```
But not:
```js
var o = {}
o.f = function() {} // o.f.name === ""
```